vulcanai2/tests/models/conftest.py

```diff
@@ -0,0 +1,205 @@
+    """Bottom multi-input network fixture."""
+    return ConvNet(
+        name='multi_input_cnn',
+        input_networks=[conv3D_net, multi_input_dnn],
+        num_classes=10,
+        config={
+            'conv_units': [
+                dict(
+                    in_channels=1,
+                    out_channels=16,
+                    kernel_size=(3, 3, 3),
+                    stride=2,
+                    dropout=0.1
+                ),
+            ],
+        },
+        device='cpu'
+    )
+@pytest.fixture(scope="module")
+def multi_input_dnn_data(conv1D_net, conv2D_net,
+                         multi_input_dnn):
+    return MultiDataset([
+        (
+            TensorDataset(
+                torch.ones([10, *conv1D_net.in_dim]),
+                torch.tensor([0, 1, 2, 3, 4, 5, 6, 7, 8, 9]).long()),
+            True, True),
+        (
+            TensorDataset(torch.ones(
+                [10, *multi_input_dnn.input_networks['conv2D_net'].in_dim])),
+            True, False)
+    ])
+
+@pytest.fixture(scope="module")
+def multi_input_cnn_data(conv3D_net, multi_input_dnn_data):
+    return MultiDataset([
+        (TensorDataset(torch.ones([10, *conv3D_net.in_dim])), True, False),
+        multi_input_dnn_data
+    ])
+
+@pytest.fixture(scope="module")
+def cnn_noclass():
+    """Create ConvNet with no prediction layer."""
+    return ConvNet(
+        name='Test_ConvNet_noclass',
+        in_dim=(1, 28, 28),
+        config={
+            'conv_units': [
+                {
+                    "in_channels": 1,
+                    "out_channels": 16,
+                    "kernel_size": (5, 5),
+                    "pool_size": 2,
+                    "stride": 2
+                },
+                {
+                    "in_channels": 16,
+                    "out_channels": 1,
+                    "kernel_size": (5, 5),
+                    "stride": 2,
+                    "padding": 2
+                }]
+        }
+    )
+
+@pytest.fixture(scope="module")
+def cnn_class():
+    """Create ConvNet with prediction layer."""
+    return ConvNet(
+        name='Test_ConvNet_class',
+        in_dim=(1, 28, 28),
+        config={
+            'conv_units': [
+                {
+                    "in_channels": 1,
+                    "out_channels": 16,
+                    "kernel_size": (5, 5),
+                    "pool_size": 2,
+                    "stride": 2
+                },
+                {
+                    "in_channels": 16,
+                    "out_channels": 1,
+                    "kernel_size": (5, 5),
+                    "stride": 2,
+                    "padding": 2
+                }]
+        },
+        num_classes=3
+    )
+
+@pytest.fixture(scope="module")
+def cnn_class_add_input_network(cnn_noclass, cnn_class):
+    """Create ConvNet with input_network added via
+    add_input_network and has a prediction layer."""
+    net = cnn_class
+    net.add_input_network(cnn_noclass)
```

Comment:
I think it adds the cnn_noclass to cnn_class which is why you might be getting the error when testing the forward on this function

Reply:
hmm I'm not quite sure if I understood `add_input_network`, lets say if a main network does not have any `input_networks` so how do I add a new input_network? I pushed a redefined example in the next commit. Let me know what am I missing.

Reply:
AHH I know the problem! So cnn_class has a dense output which you are trying to feed into a conv net. It's an edge case that I let fail since I think it doesn't make much sense for that sort of scenario (e.g. fully connected -> cnn). If you disagree, then it would not be a difficult fix; you would just need to pass a cast_shape of the receiving network to cast the dense input to that. @priyatharsan @joemehltretter @nirtiac I'd like to get your input on this.

Reply:
I think its good to have that ability to convert FC to CNN, as we already have the ability to convert [FC + some other conv input] to a CNN.

Reply:
However, I'm confused in this case `cnn_class.add_input_network(cnn_noclass)`, I'm adding `cnn_noclass` (which has a conv output) as an input_network to the `cnn_class` right?

Reply:
@priyatharsan the issue is with information loss. dense + conv 1D inputted to conv 2D doesn't make much sense since you're not getting any additional benefit of padding a bunch of zeros in the extra dimension. Really the input network should have a size of the max(list_input_networks_output_size) which is what is happening now. Similarly, FC -> Conv doesn't really add much benefit since there's no spatial information in FC. Does that make sense? Regarding debugging, I would see what the cnn_noclass output shape is